# Used cars from Ebay Kleinanzeigen

Ebay Kleinanzeigen is a free online classified German Market place.
The original dataset was scraped and uploaded to Kaggle by user [Ordesleka](https://www.kaggle.com/orgesleka).</p><p> We sampled 50,000 data points from the original dataset, which is what we will be working with.</p><p> We also dirtied the dataset a bit to ensure a more close resemblance to what's expected when scraping data from sites.</p>

In [1]:
#import libraries
import numpy as np
import pandas as pd

# create readable csv
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
# show the autos dataset
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
# get information about the dataset
print(autos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

We can see from the information above, we have 50,000 entries and 20 columns. <p> We can also tell there are some null values in our dataset. The columns with null values are `vehicleType`, `gearbox`, `model`, `fuelType` and `notRepairedDamage`. However these null values account for less than 20%.<p> The 20 columns are as follow:
    
| Column               | Description                                                      |
| :------------------: |:--------------------------------------------------------------- |
| dateCrawled          | When this ad was first crawled. All field-values are taken from this date. |
| name                 | Name of the car.                                                 |
| seller               | Whether the seller is private or a dealer.                       |
| offerType            | The type of listing.                                             |
| price                | The price on the ad to sell the car.                              |
| abtest               | Whether the listing is included in an A/B test.                   |
| vehicleType          | The vehicle Type.                                                |
| yearOfRegistration   | The year in which the car was first registered.                   |
| gearbox              | The transmission type.                                           |
| powerPS              | The power of the car in PS.                                       |
| model                | The car model name.                                               |
| odometer             | How many kilometers the car has driven.                           |
| monthOfRegistration  | The month in which the car was first registered.                  |
| fuelType             | What type of fuel the car uses.                                   |
| brand                | The brand of the car.                                             |
| notRepairedDamage    | If the car has a damage which is not yet repaired.                 |
| dateCreated          | The date on which the eBay listing was created.                   |
| nrOfPictures         | The number of pictures in the ad.                                 |
| postalCode           | The postal code for the location of the vehicle.                   |
| lastSeenOnline       | When the crawler saw this ad last online.  

In [4]:
# see first 5 rows of our dataset
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [5]:
# show an array of the column names
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

Some of the columns use [camelcase](https://en.wikipedia.org/wiki/Camel_case) instead of the preferred [snakecase](https://en.wikipedia.org/wiki/Snake_case) peferred by Python.

In [6]:
# update the column names to be more uniform
new_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_of_pictures', 'zip_code',
       'last_seen']

autos.columns = new_columns

In [7]:
# look at first few rows to make sure the changes took place
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00                0   
1         bmw              nein  2016-04-04 00:00:00                0   
2  volkswagen              nein  2016-03-26 00:00:00                0   
3       smart              nein  2016-03-12 00:00:00                0   
4        ford              nein  2016-04-01 00:00:00                0   

   zip_code            last_seen  
0     79588  2016-04-06 06:45:54  
1     71034  2016-04-06 14:45:08  
2     35394  2016-04-06 20:15:37  
3     33729  2016-03-15 03:16:28  
4     39218  2016-04-01 14:38:50

We have now successfully updated the [camelcase](https://en.wikipedia.org/wiki/Camel_case) to Python-preferred [snakcase](https://en.wikipedia.org/wiki/Snake_case).

These columns include but not limited to 
<br>`vehicleType` to `vehicle_type`<br>
`yearOfRegistration` to `registration_year`  
`powerPS` to `power_ps`  
`monthOfRegistration` to `registration_month`.  

## Additional cleaning and Exploration

This section we will be doing some basic exploration to determine what other cleaning task need to be done. Initially we will look for:

* Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.


* Examples of numeric data that are stored as text which can be cleaned and converted.

In [8]:
# summary statistics for all regardless of data type
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_of_pictures      zip_code  \
count                 50000          50000.0  50000.000000   
unique                   76              NaN           NaN   
top     2016-04-03 00:00:00              NaN           NaN   
freq                   1946              NaN           NaN   
mean                    NaN              0.0  50813.627300   
std                     NaN              0.0  25779.747957   
min                     NaN              0.0   1067.000000   
25%                     NaN              0.0  30451.000000   
50%                     NaN              0.0  49577.000000   
75%                     NaN              0.0  71540.000000   
max                     NaN              0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

The summary above shows us the following:
    
  * `seller` and `offer_type` are almost all the same, so we can drop them sense they want have any real value in our analysis.
  
  * `price` and `odometer` con columns contain text instead of numerical values, and we will need to change that. We can tell from the $ in price and the km on odemeter.
  
  * `registration_year` shows a minimum of 1000 and a maximum of 9999. We need to investigate this further and determine if we have any outliers or incorrect data.

In [9]:
# check to see the data types
print(autos["price"].dtype)
print(autos["odometer"].dtype)

object
object


Our initial thoughts about `price` and `odometer` were stored as strings was confirmed.

In [10]:
# change price and odometer to numerical values
autos["price"] = autos["price"].str.replace('$', '').str.replace(',', '').astype(int)
autos["odometer"] = autos["odometer"].str.replace('km', '').str.replace(',', '').astype(int)

# rename the odometer column
autos.rename(columns={"odometer": "odometer_km"}, inplace=True)

print(autos["price"].dtype)
print(autos["odometer_km"].dtype)

int64
int64


The code above changed the `price` and `odometer` columns to numerical values. We also changed the `odometer` column to `odometer_km`.

## Continue to explore the data:

* We will continue exploring looking for any [outliers](https://www.techopedia.com/definition/32511/outlier) that we may want to remove.

In [11]:
# find the amount of unique values
print("The number of unique values for price are:", autos["price"].unique().shape)
print("The number of unique values in odometer_km are:", autos["odometer_km"].unique().shape)

The number of unique values for price are: (2357,)
The number of unique values in odometer_km are: (13,)


In [12]:
# look at min and max
print(autos["price"].describe())
print("\n")
print(autos["odometer_km"].describe())

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


The output above shows us in `price` that 75% is 7.5 million or less, Whereas the max is 10 million, which is an indication of an outlier.

In [13]:
# get the highest unique price values and their counts
print(autos["price"].value_counts().sort_index(ascending=False).head(10))
print("\n")
# get the highest unique odometer values and their counts
print(autos["odometer_km"].value_counts().sort_index(ascending=False).head(10))

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
Name: odometer_km, dtype: int64


In [14]:
# get the lowest unique price value and their counts
print(autos["price"].value_counts().sort_index(ascending=True).head(10))
print("\n")
# get the lowest unique odometer value and their counts
print(autos["odometer_km"].value_counts().sort_index(ascending=True).head(10))


0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price, dtype: int64


5000      967
10000     264
20000     784
30000     789
40000     819
50000    1027
60000    1164
70000    1230
80000    1436
90000    1757
Name: odometer_km, dtype: int64


In [15]:
# remove the outliers in price anything above 1 million
autos = autos[autos["price"] < 1000000]

In [16]:
# display the updated statistics
autos["price"].describe()

count     49989.000000
mean       5781.194763
std       11833.263624
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      999999.000000
Name: price, dtype: float64

* The updated statistics for the `price` column after removing the outliers show us that the average price of a car in the dataset is now `5781`, which is significantly lower than the previous mean of `9840`. 
* The standard deviation has also decreased from `481104` to `11833`, indicating that the prices are now more tightly clustered around the mean. 
* Overall, these statistics suggest that removing the outliers has resulted in a more representative distribution of car prices in the dataset.

## Explore the data columns

In this section we will explore the date columns. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary.

|                |                                                     |
| :------------------: |:--------------------------------------------------------------- |
| date_crawled          | added by the crawler |
| last_seen            | added by the crawler                                |
| ad_created           | from the website          |
| registration_month           | from the website                       |
| registration_year                | from the website  

In [17]:
# review the date columns
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

* The `data_crawled`,  `last_seen`, `ad_created` columns are all string values. Since these columns are represented by string, We need to convert them into numerical representations to better understand them quantitatively"


* The `registration_month` and `registration_year` are already in numeric values.

In [18]:
# extract just the dates (excluding timestamp) 
# and print first 5 rows
print(autos['date_crawled'].str[:10].head())
print('\n')
print(autos['last_seen'].str[:10].head())
print('\n')
print(autos['ad_created'].str[:10].head())

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: date_crawled, dtype: object


0    2016-04-06
1    2016-04-06
2    2016-04-06
3    2016-03-15
4    2016-04-01
Name: last_seen, dtype: object


0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: ad_created, dtype: object


In [19]:
# calulate the distribution as percentages
# rank date by ascending order
print(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.025386
2016-03-06    0.013943
2016-03-07    0.035968
2016-03-08    0.033267
2016-03-09    0.033207
2016-03-10    0.032127
2016-03-11    0.032487
2016-03-12    0.036768
2016-03-13    0.015563
2016-03-14    0.036628
2016-03-15    0.033987
2016-03-16    0.029506
2016-03-17    0.031527
2016-03-18    0.013063
2016-03-19    0.034908
2016-03-20    0.037828
2016-03-21    0.037508
2016-03-22    0.032907
2016-03-23    0.032387
2016-03-24    0.029106
2016-03-25    0.031747
2016-03-26    0.032487
2016-03-27    0.031047
2016-03-28    0.034848
2016-03-29    0.034168
2016-03-30    0.033627
2016-03-31    0.031907
2016-04-01    0.033807
2016-04-02    0.035408
2016-04-03    0.038689
2016-04-04    0.036488
2016-04-05    0.013103
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64


This output shows the distribution of when the listings were first crawled by the web crawler, from `2016-03-05` to `2016-04-07`. It tells us the percentage of listings crawled on each date, with the dates sorted in ascending order. 

For example, on `2016-03-05`, `2.54%` of the total listings were crawled, and on `2016-04-07`, only `0.14%` of the total listings were crawled.

In [20]:
# calulate the distribution as percentages
# rank date by ascending order
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005361
2016-03-08    0.007582
2016-03-09    0.009842
2016-03-10    0.010762
2016-03-11    0.012523
2016-03-12    0.023805
2016-03-13    0.008982
2016-03-14    0.012803
2016-03-15    0.015883
2016-03-16    0.016444
2016-03-17    0.027926
2016-03-18    0.007422
2016-03-19    0.015743
2016-03-20    0.020705
2016-03-21    0.020725
2016-03-22    0.021585
2016-03-23    0.018584
2016-03-24    0.019564
2016-03-25    0.019204
2016-03-26    0.016964
2016-03-27    0.016024
2016-03-28    0.020865
2016-03-29    0.022345
2016-03-30    0.024845
2016-03-31    0.023825
2016-04-01    0.023105
2016-04-02    0.024885
2016-04-03    0.025366
2016-04-04    0.024625
2016-04-05    0.124267
2016-04-06    0.220989
2016-04-07    0.130949
Name: last_seen, dtype: float64


This output shows the distribution of dates in the `last_seen` column, which represents the last date the crawler saw the ad online before it was removed. It is important to note that the "last_seen" column represents the date the crawler saw the ad online and not the date it was sold or removed. The percentage value indicates the proportion of ads that were last seen on a particular date, ranked in ascending order.

We can observe that the majority of the ads were last seen in April, with a sudden surge in the last three days of crawling. This could be attributed to the removal of a significant number of ads around that time. Additionally, the dates in March have a lower proportion of ads seen compared to April. This suggests that there was a higher turnover rate of ads in March, which were replaced by new ads in April.

In [21]:
# calulate the distribution as percentages
# rank date by ascending order
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
                ...   
2016-04-03    0.038929
2016-04-04    0.036848
2016-04-05    0.011843
2016-04-06    0.003261
2016-04-07    0.001280
Name: ad_created, Length: 76, dtype: float64


This output shows the distribution of dates in the `ad_created` column, which represents the date the ad was created on the website. The percentage value indicates the proportion of ads that were created on a particular date, ranked in ascending order.

We can observe that the majority of the ads were created in March and April of 2016, with a smaller number of ads created in the months leading up to that period. It is also worth noting that there are some ads that were created as early as June 2015, but they make up a very small proportion of the total number of ads.

In [22]:
# distribution of registration year
autos["registration_year"].describe()

count    49989.000000
mean      2005.074916
std        105.724083
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The statistics on the `registration_year` seem to have a few outliers. We can see it has a minimum of 1000 and a maximum of 9999, which doesn't make much sense for a registration year.

We need to further explore this data to understand whats going on.

In [23]:
# get the unique value count percentages
print(autos["registration_year"].value_counts(normalize=True))
print()
# get the unique value counts by date
print(autos["registration_year"].value_counts())

2000    0.067095
2005    0.060313
1999    0.059973
2004    0.054752
2003    0.054552
          ...   
1931    0.000020
1929    0.000020
1001    0.000020
9996    0.000020
1952    0.000020
Name: registration_year, Length: 97, dtype: float64

2000    3354
2005    3015
1999    2998
2004    2737
2003    2727
        ... 
1931       1
1929       1
1001       1
9996       1
1952       1
Name: registration_year, Length: 97, dtype: int64


Since it's unlikely that any cars from the 1900s are still registered today, we can consider those registration years as outliers and remove them from our dataset. Additionally, we should also remove the years '1001' and '9996' since they do not make any sense for car registration years.

We can also see that the cars in the 1900s only make up `0.002%` of our cars.

In [30]:
# remove the outliers - anything less than the year 1999
autos = autos[autos["registration_year"] < 1999]

# check the remaining distribution
autos["registration_year"].value_counts(normalize=True)

1998    0.225480
1997    0.186414
1996    0.132733
1995    0.120599
1994    0.060667
          ...   
1948    0.000092
1111    0.000092
1927    0.000092
1943    0.000092
1000    0.000092
Name: registration_year, Length: 65, dtype: float64

Now than we have removed the outliers, you can see that majority of the cars were registered between the years `1995` - `1998`. The year `1998` being the highest year with around `22.5%` and `1994` being the lowest with `6%`.

## Continue exploring the data:

* Explore the unique values in the brand column, and decide which brands to aggregate by.

In [25]:
# see what car brands distribute the most to our dataset
print(autos["brand"].value_counts(normalize=True))

volkswagen        0.274566
opel              0.125471
bmw               0.116831
mercedes_benz     0.096424
audi              0.076845
ford              0.056899
renault           0.041088
sonstige_autos    0.021325
fiat              0.020498
peugeot           0.015994
nissan            0.014615
mazda             0.014431
volvo             0.013788
honda             0.013053
seat              0.012409
toyota            0.009376
mitsubishi        0.009100
porsche           0.007446
trabant           0.006894
citroen           0.006159
chevrolet         0.005975
suzuki            0.004872
alfa_romeo        0.004688
chrysler          0.004596
subaru            0.003493
jeep              0.002941
saab              0.002758
skoda             0.002758
daihatsu          0.002390
jaguar            0.002022
rover             0.001838
lancia            0.001563
hyundai           0.001379
land_rover        0.001379
kia               0.001011
lada              0.001011
mini              0.000919
d

Lets explore the top 10 brands.

Those brands are as follows:

| Brands      | Distribution(Percentage)|
:-------------|-------------------------|
|volkswagen   |     0.274566            |
|opel         |     0.125471            |
|bmw          |      0.116831           |
|mercedes_benz|     0.096424            
|audi         |     0.076845
|ford         |     0.056899
|renault      |     0.041088
|sonstige_autos|    0.021325
|fiat          |    0.020498
|peugeot       |    0.015994

In [26]:
# create list of the brands we want
unique_brands = ["volkswagen", "opel", "bmw", "mercedes_benz", "audi",
                "ford", "renault", "sonstige_autos", "fiat" ]

# store the mean/average of those brands
brand_mean = {}

# iterate over our unique_brands
for brand in unique_brands:
    
    # create subset for our brands
    brand_subset = autos[autos["brand"] == brand]
    
    # calculate the mean
    mean_price = brand_subset["price"].mean()
    
    # assign brand to mean_price
    brand_mean[brand] = round(mean_price, 3)
 
# print out the brand_mean
print(brand_mean)

{'volkswagen': 2439.823, 'opel': 970.118, 'bmw': 2429.78, 'mercedes_benz': 4394.45, 'audi': 1697.929, 'ford': 2729.155, 'renault': 951.128, 'sonstige_autos': 9229.151, 'fiat': 1755.722}


The brand data shows that Volkswagen is the most common brand in the dataset, with a distribution of `27.5%`. The top 10 brands by distribution make up over `75%` of the total values, suggesting that they are the most popular brands among the car listings. I selected these top 10 brands to aggregate on because they provide a representative sample of the most common brands in the dataset. By calculating the average car prices for each brand, we can gain insights into the relative prices of different car brands. Notably, `sonstige_autos` has the highest average car price of `$9,229.15`, while `renault` has the lowest average car price of `$951.12`. These findings could be useful for car buyers or sellers who are interested in understanding the average prices of different brands.

In [27]:
# create a series constructor
bmp_series = pd.Series(brand_mean)
print(bmp_series)

volkswagen        2439.823
opel               970.118
bmw               2429.780
mercedes_benz     4394.450
audi              1697.929
ford              2729.155
renault            951.128
sonstige_autos    9229.151
fiat              1755.722
dtype: float64


In [33]:
# combine the data from both series objects into single dataframe
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df

mean_price
volkswagen        2439.823
opel               970.118
bmw               2429.780
mercedes_benz     4394.450
audi              1697.929
ford              2729.155
renault            951.128
sonstige_autos    9229.151
fiat              1755.722

In [36]:
# create list of the brands we want
top_brands = ["volkswagen", "opel", "bmw", "mercedes_benz", "audi",
                "ford", "renault", "sonstige_autos", "fiat" ]

# store the mean/average of those brands
brand_mean_price = {}
brand_mean_mileage = {}

# iterate over our unique_brands
for brand in top_brands:
    
    # create subset for our brands
    brand_subset = autos[autos["brand"] == brand]
    
    # calculate the mean
    mean_mileage = brand_subset["odometer_km"].mean()
    mean_price = brand_subset["price"].mean()
    
    # assign brand to mean_price and mean_mileage
    brand_mean_price[brand] = round(mean_price, 3)
    brand_mean_mileage[brand] = mean_mileage

In [ ]:
# create series for mean price and mean mileage
price_series = pd.Series(brand_mean_price)
mileage_series = pd.Series(brand_mean_mileage)

In [ ]:
#create a tabular dataframe for mean_price and mean_mileage
mean_df = pd.DataFrame({"mean_price": price_series, "mean_mileage": mileage_series})
mean_df.sort_values(["mean_price"], ascending=False)

From the data, we can see that there is significant variation in both the mean price and mean mileage among the top car brands. 

- The brand with the highest mean price is "Sonstige Autos" with a value of 9229.151, while the brand with the lowest mean price is "Renault" with a value of 951.128. 

- On the other hand, the brand with the highest mean mileage is "Audi" with a value of 145023.923, while the brand with the lowest mean mileage is "Sonstige Autos" with a value of 77025.862. 

It is interesting to note that "Sonstige Autos" has the highest mean price but the lowest mean mileage, indicating that this brand may be associated with high-end luxury vehicles that are not driven as frequently. Overall, this data provides valuable insights into the market trends for these popular car brands.